### PR TITLE
feat(eventing): state change events for nexus and replica

### PR DESCRIPTION
**Pull request by datacore-vvarakantham**
Added state change events for nexus and replica.

### CODE REVIEW

File: io-engine/src/bdev/nexus/nexus_bdev.rs [hunk 1]
```diff
@@ -851,7 +858,7 @@ impl<'n> Nexus<'n> {
             let name = self.name.clone();
 
             // After calling unregister_bdev_async(), Nexus is gone.
-            let evt = self.event(EventAction::Delete);
+            let evt = Event::event(&(*self), EventAction::Delete);
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev.rs]: Why this change is needed?

File: io-engine/src/bdev/nexus/nexus_bdev.rs [hunk 2]
```diff
@@ -909,6 +916,12 @@ impl<'n> Nexus<'n> {
             // Reset operation is allowed only when the Nexus is Open state
             NexusState::Open => {
                 *state = NexusState::Reconfiguring;
+                EventWithMeta::event(
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev.rs]: Don't think we need such fine details, state changing to showdown should be enough?

File: io-engine/src/bdev/nexus/nexus_bdev.rs (hunk 1 above)
Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/nexus_bdev.rs]: I had to use the event() function from two different traits Event and EventWithMeta in the same module. I added this to use event() fn from Event trait.

File: io-engine/src/bdev/nexus/nexus_bdev.rs (hunk 2 above)
Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/nexus_bdev.rs]: Yes the events list is filled with many state change events as we are having event for every state change. Other state change events are not needed for nexus other than shutdown? Or should I just remove reconfiguring event?

Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/nexus_bdev.rs]: Removed nexus state change events for reconfiguring and open states.

File: io-engine/src/bdev/nexus/nexus_bdev.rs (hunk 1 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_bdev.rs]: Perhaps, deref might be neater, but just a nit:
```suggestion
            let evt = Event::event(self.deref(), EventAction::Delete);
```

File: io-engine/src/eventing/nexus_events.rs [hunk 3]
```diff
@@ -50,6 +53,16 @@ impl<'n> EventMetaGen for NexusChild<'n> {
     }
 }
 
+pub(crate) fn state_change_event_meta(
```
Comment (tiagolobocastro) [inline io-engine/src/eventing/nexus_events.rs:59]: pls add doc comment

File: io-engine/src/eventing/replica_events.rs [hunk 4]
```diff
@@ -32,3 +36,28 @@ impl Event for Lvol {
         }
     }
 }
+
+pub(crate) fn state_change_event_meta(
```
Comment (tiagolobocastro) [inline io-engine/src/eventing/replica_events.rs:41]: doc comment here

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/eventing/replica_events.rs (hunk 4 above)
Comment (datacore-vvarakantham) [reply io-engine/src/eventing/replica_events.rs:41]: Added the doc comment.

**Comment by hrudaya21** [comment]
@tiagolobocastro Do We need to consider Subsystem pause/resume event and nexus init/reconfigure event ?

**Comment by tiagolobocastro** [comment]
> @tiagolobocastro Do We need to consider Subsystem pause/resume event and nexus init/reconfigure event ?

You mean as events? Yes those would be very useful as well!

**Comment by hrudaya21** [comment]
> > @tiagolobocastro Do We need to consider Subsystem pause/resume event and nexus init/reconfigure event ?
> 
> You mean as events? Yes those would be very useful as well!

I think, We can handle these scenario in this PR, as We already handling nexus events here. @datacore-vvarakantham

**Comment by datacore-vvarakantham** [comment]
> > > @tiagolobocastro Do We need to consider Subsystem pause/resume event and nexus init/reconfigure event ?
> > 
> > 
> > You mean as events? Yes those would be very useful as well!
> 
> I think, We can handle these scenario in this PR, as We already handling nexus events here. @datacore-vvarakantham

Sure, I can add those as events.

File: io-engine/src/bdev/nexus/nexus_bdev.rs [hunk 5]
```diff
@@ -1017,8 +1052,42 @@ impl<'n> Nexus<'n> {
...
+        .generate();
+        let start_time = std::time::Instant::now();
+        let result = self.as_mut().io_subsystem_mut().suspend().await;
+        match result {
+            Ok(_) => {
+                EventWithMeta::event(
+                    self.deref(),
+                    EventAction::SubsystemPause,
+                    subsystem_pause_event_meta(
+                        "completed",
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev.rs]: "completed" or "failed" seems to be unnecessary the way it;s used here, since we can infer that from the error.
I think including the actual subsystem status would be better here.

Comment (datacore-vvarakantham) [reply io-engine/src/bdev/nexus/nexus_bdev.rs]: Yes. I added NexusPauseState now.

**Review by sinhaashish — APPROVED** [review]

**Comment by datacore-vvarakantham** [comment]
bors merge
